clang/runtime/dpct-rt/include/sparse_utils.hpp.inc

```diff
@@ -78,6 +79,241 @@ private:
   oneapi::mkl::index_base _index_base = oneapi::mkl::index_base::zero;
 };
 // DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|csrmv|dpct::sparse
+// DPCT_DEPENDENCY_BEGIN
+// SparseUtils|matrix_info
+// Util|DataType
+// LibCommonUtils|get_value
+// LibCommonUtils|get_memory
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+/// Computes a CSR format sparse matrix-dense vector product.
+/// y = alpha * op(A) * x + beta * y
+/// \param [in] queue The queue where the routine should be executed.
+/// \param [in] trans The operation applied to the matrix A.
+/// \param [in] num_rows Number of rows of the matrix A.
+/// \param [in] num_cols Number of columns of the matrix A.
+/// \param [in] alpha Scaling factor for the matrix A.
+/// \param [in] info Matrix info of the matrix A.
+/// \param [in] val An array containing the non-zero elements of the matrix A.
+/// \param [in] row_ptr An array of length \p num_rows + 1.
+/// \param [in] col_ind An array containing the column indices in index-based
+/// numbering.
+/// \param [in] x Data of the vector x.
+/// \param [in] beta Scaling factor for the vector x.
+/// \param [in, out] y Data of the vector y.
+template <typename T>
+void csrmv(sycl::queue &queue, oneapi::mkl::transpose trans, int num_rows,
+           int num_cols, const T *alpha,
+           const std::shared_ptr<matrix_info> info, const T *val,
+           const int *row_ptr, const int *col_ind, const T *x, const T *beta,
+           T *y) {
```

Comment:
I don't see any maintenance of event patterns here,  are we always mapping to an inline queue  or if it is an asynchronous queue, are there input dependencies that could be passed in and out ?

I will add suggestions below that highlight the internal consistency of operations as if it were an asynchronous queue.  If it is always inline queue, then feel free to ignore them :)

Reply:
All the queues in the migrated code has the `in_order` property if using USM mode.

Reply:
ok so in_order essentially removes the need for passing event to event, but i still recommend changing from matrix_handle_t * type to  matrix_handle_t type.